### ISSUE TITLE
change request to kobo form sharing notification workflow to detect when form changed from `deployed` to `archived`

### ISSUE THREAD

**Issue by aleksa-krolls** [OpenFn/ConSoSci#214]
## Background, context, and business value
This is a change request to #206 where OpenFn...
1. Queries forms from Kobo where `"deployment__active": true` - [see example payload](https://github.com/OpenFn/ConSoSci/blob/master/sample_data/sample-asset-list.json)
2. Updates [this GoogleSheet](https://docs.google.com/spreadsheets/d/1s7K3kxzm5AlpwiALattyc7D9_aIyqWmo2ubcQIUlqlY/edit#gid=1559623602) with details about the deployed form
3. Creates a Task in Asana if a _new_ form is added to this^ sheet to alert Diane that a new data collection form has been created by her field team

**Note that WCS is still on platform v1**. 

## The specific request, in as few words as possible
Now the client is wondering if it would be possible to _also_ "upsert" the rows in the GoogleSheet to keep the `status` and `kobo_date_modified` up-to-date. We will need [this upsertRows()](https://app.zenhub.com/workspaces/implementation-engineering-600030af88f2bb00155d46a5/issues/gh/openfn/adaptors/510) adaptors issue to deliver this. 

There are scenarios where a field team will deploy a new Kobo form, which would result in a new row in the sheet. However, they might then later _archive_ the form (changing the `deployment__active` from `true` to `false`). How could we design this workflow to then update the `status` and `kobo_date_modified` columns accordingly? And then alert the client via an Asana task when the form is archived? 

- see [workflow v2 diagram](https://lucid.app/lucidchart/346b8e5c-6fb6-4a33-9d02-53e5059bd698/edit?beaconFlowId=084EDC490DE90DA7&invitationId=inv_d1431bce-05ae-4005-9b6a-9c279141a3a3&page=QNkDcyLzfaKX#) for change requests/additions
- see [pink sheets](https://docs.google.com/spreadsheets/d/1dEjcKwLWLFsROsGXefWsryHRi0pvUGGTLndFwhifWU4/edit#gid=1270931487) for related mapping spec changes

**Do you think this should be new jobs that run in parallel? Or edits to the existing jobs?** 

### API docs
- Kobo: https://kf.kobotoolbox.org/api/v2/assets/
- GoogleSheets: https://developers.google.com/sheets/api/guides/concepts
- Asana: https://developers.asana.com/reference/createtask

## jobs (expression.js)
Please create a new branch off of `master` --> see existing workflow in the [form-sharing folder](https://github.com/OpenFn/ConSoSci/tree/master/form-sharing). 

- `1-getNewKoboForms.js`  (adaptor: kobo)
- `2-getSheetsList.js` (adaptor: googlesheets)
- `3-updateSheetsList.js` (adaptor: googlesheets)
- `4-upsertAsanaTask.js` (adaptor: asana)

### Job 1: Kobo 
- configuration: see LP `OpenFn Kobo Toolbox (for WCS data access!)`
- data: {}

### Job 2 & 3: GoogleSheets 
- configuration: For now use your Google user - you have access to [this sheet](https://docs.google.com/spreadsheets/d/1s7K3kxzm5AlpwiALattyc7D9_aIyqWmo2ubcQIUlqlY/edit#gid=0)
- data: { new kobo forms detected}

### Job 4: Asana 
- configuration: For now use your Asana user - you have access to [this project](https://app.asana.com/0/1198901998266253/1205105885958289)
- data: { new kobo forms detected}

## output
If a Kobo form was previously saved in the GoogleSheet but in Kobo `deployment__active: false` is now detected...
1. Upsert the row in the GoogleSheet to change `status` from `deployed` --> `archived` and update `kobo_date_modified` to reflect the latest date seen in Kobo
2. New task created in Asana that tells the admin that a form has been archived

When ready to test, we'll update the jobs [in this V1 project](https://v1.openfn.org/projects/consosci-training/jobs) - search for `FS` to see related jobs

## toggl
Log your time spent on this project: [WCS Support 2023/24](https://track.toggl.com/3335160/projects/195655331/tasks)

**Comment by mtuchi** [comment]
>alert the client via an Asana task when the form is archived?

@aleksa-krolls Can you tell me more about alerting the client ?, Is this like an Asana notification email ?

**Comment by aleksa-krolls** [comment]
@mtuchi all you need to do is upsert the Task in Asana. (Native Asana notifications will alert the user once this task is upserted.)

**Comment by mtuchi** [comment]
@aleksa-krolls i have opened the PR for this issue See #217, I suggest we sync during QA. We have decided not to implement the `upsertRows()` helper function, [See joe's feedback](https://github.com/OpenFn/adaptors/issues/510#issuecomment-2124223476)

##### Here are QA Notes 

1. Sync form which match keyword list `['price', 'prix', 'bns', 'nrgt']` and have `deployment__active: true`
2. Check in google sheet if all forms are synced.
3. Then change the status of one of the already synced form to `deployment__active: false`
4. Then re-run the workflow again
5. Check column `status` and `kobo_date_modified` to see the changes

**Comment by aleksa-krolls** [comment]
thanks @mtuchi! Let's chat through what you've implemented & Joe's feedback in standup tomorrow to align on next steps
